### PR TITLE
Improvements on QosPolicies.h [7892]

### LINKED ISSUE #1053: Compiler warning TemplateDataQosPolicy deprecated copy [7877]
This happens with newer gcc on Ubuntu Focal: see e.g. http://build.ros2.org/view/Fci/job/Fci__nightly-fastrtps_ubuntu_focal_amd64/51/warningsResult/

`ReaderQos.h:35`:

```
implicitly-declared  ‘eprosima::fastrtps::TemplateDataQosPolicy&  eprosima::fastrtps::TemplateDataQosPolicy::operator=(const  eprosima::fastrtps::TemplateDataQosPolicy&)’ is deprecated  [-Wdeprecated-copy]
```

### PR DESCRIPTION
**Pull request by MiguelCompany**
This PR will fix #1053 as well as introduce some improvements on the whole `QosPolicies.h` file

### CODE REVIEW

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1546/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1318/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1544/)

File: include/fastrtps/qos/QosPolicies.h [hunk 1]
```diff
@@ -802,6 +803,7 @@ class GenericDataQosPolicy : public Parameter_t, public QosPolicy, public Resour
             const std::vector<rtps::octet>& vec)
     {
         assign(vec.begin(), vec.end());
+        length = (size() + 7) & ~3;
     }
```
Comment (IkerLuengo) [inline include/fastrtps/qos/QosPolicies.h]: set `hasChanged` to `true`

File: include/fastrtps/qos/QosPolicies.h [hunk 2]
```diff
@@ -821,6 +823,7 @@ class GenericDataQosPolicy : public Parameter_t, public QosPolicy, public Resour
...
```
Comment (IkerLuengo) [inline include/fastrtps/qos/QosPolicies.h]: set hasChanged to true

**Comment by richiware** [comment]
Build status:
* Linux [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Linux/1560/)
* Mac [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Mac/1332/)
* Windows [Build Status](http://jenkins.eprosima.com:8080/job/FastRTPS%20Manual%20Windows/1558/)

**Pull request merged**